File: DSAlgorithms/Leetcode/ts/src/stacks/155-min_stack.ts
class MinStack {
    a_stack: number[] = [];
    min_stack: number[] = [];
    constructor() {
        this.a_stack = [];
        this.min_stack = [];
    }

    push(val: number): void {
        this.a_stack.push(val);
        let new_min = 0;
        if(this.min_stack.length > 0)
            new_min = Math.min(val, this.min_stack[this.min_stack.length - 1])
        else 
            new_min = val;
        this.min_stack.push(new_min);
    }

    pop(): void {
        this.a_stack.pop();
        this.min_stack.pop();
    }

    top(): number {
        return this.a_stack[this.a_stack.length - 1];

    }

    getMin(): number {
        return this.min_stack[this.min_stack.length - 1];
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: DSAlgorithms/Leetcode/c++/stacks/155-min_stack.cpp
class MinStack {
    std::stack<int> a_stack;
    std::stack<int> min_stack;
public:
    MinStack() {
       a_stack = {};
       min_stack = {};
    }
    
    void push(int val) {
        a_stack.push(val);
        int new_min = 0;
        if(!min_stack.empty())
            new_min = std::min(val, min_stack.top());
        else new_min = val;
        min_stack.push(new_min);
    }
    
    void pop() {
       a_stack.pop();
       min_stack.pop();
    }
    
    int top() {
       return a_stack.top();
    }
    
    int getMin() {
       return min_stack.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */